### PR TITLE
♿ Announce story discovery dialog for screen readers

### LINKED ISSUE #32517: amp-story: Status messages are not explicitly announced by assistive technologies
# Status messages are not explicitly announced by assistive technologies

## WCAG Level
Level AA

## Priority
Medium

## Pages/screens/components affected
* [AMP Story Player - AMP Story 360](https://www.gstatic.com/amphtml/stamp/qa/amp-story-360.html)

## Description
When status messages/notifications appear on a page dynamically, without automatically receiving keyboard focus, it is important that these messages also be communicated correctly to assistive technologies.

The instruction overlay for the 360 AMP Story

In [AMP Story Player - AMP Story 360](https://www.gstatic.com/amphtml/stamp/qa/amp-story-360.html) an instruction overlay is shown on the second page, which disappears after a few seconds. This overlay is not announced by assistive technologies.

## User impact
People using assistive technologies such as screen readers and screen magnification tools may not be aware when status messages are injected into the page outside of their current area of focus. For example, if a confirmation message only appears visually (without being explicitly announced) as a "toast" style notification, or even as a permanently visible new notification block in a specific part of the screen that is not currently in focus, screen reader users may not even be aware that a confirmation message appeared at all.

## Required solution
Whenever status messages or notifications are injected in the page, and these status messages do not receive focus (which is usually not appropriate, since this moves the user's context and interrupts their current activity), make sure that the messages are coded in such a way that they are announced by assistive technologies. For web-based content, this typically means identifying the status message as an [ARIA live region](https://www.w3.org/TR/wai-aria-1.1/#attrs_liveregions) - by either using a [live region `role`](https://www.w3.org/TR/wai-aria-1.1/#live_region_roles), or adding an [`aria-live` attribute](https://www.w3.org/TR/wai-aria-1.1/#aria-live).

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

## Implementation guidance

The current overlay is already set up as an ARIA live region, using `aria-live="polite"`. However, the actual text content "Move device to explore" is hidden using `aria-hidden="true"`.

```
<div class="i-amphtml-story-360-discovery" aria-live="polite">
  <div class="i-amphtml-story-360-discovery-animation"></div>
  <span class="i-amphtml-story-360-discovery-text" aria-hidden="true">Move device to explore</span>
</div>
```

Remove the `aria-hidden="true"` from the `<span>`, which should ensure that the overlay is correctly announced by assistive technologies when it appears.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Turn on your screen reader, and open a page that includes a status message.
2. Activate the component that triggers the status message.
3. Make sure that screen readers announce appropriate feedback when the status message visibly appears on the page.

### Definition of done
Complete all of these tasks before closing this issue or indicating it is ready for retest:
* All issues identified within the test sample have been resolved.
* The rest of the website has been tested for the same issue.
* All issues identified throughout the website have been resolved or filed as new issues.

## Related standards
[WCAG 2.1 Success Criterion 4.1.3 Status Messages (Level A)](https://www.w3.org/TR/WCAG21/#status-messages)

## More information
* [W3C - Understanding Success Criterion 4.1.3: Status Messages](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html)
* [Knowbility - Exploring WCAG 2.1: 4.1.3 Status Messages](https://knowbility.org/blog/2018/WCAG21-413StatusMessages/)
* [MDN web docs - ARIA live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)

## Test data
Test date: January 2021

### PR DESCRIPTION
**Pull request by nickFalcone**
Per #32517, discovery text should be announced for screen readers.  In this case, the text is "Move device to explore" (gyroscope experience, [example](https://www.gstatic.com/amphtml/stamp/qa/amp-story-360.html#page=page2)).

- Removed `aria-hidden="true"` to allow screen reader to read discovery dialog
- Tested with JAWS 21, "Move device to explore" is read aloud

Consideration: The gyroscope experience, by its nature, may not be useable by people who are blind.  However, the notification message could still be helpful for people with limited vision. 

The discovery dialog is displayed initially, then hidden via [fadeOut animation](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story-360/0.1/amp-story-360.css#L89).  

Resolves #32517

/cc @Enriqe

### CODE REVIEW

**Review by Enriqe — APPROVED** [review]
This LGTM, thanks @nickFalcone !

/cc @processprocess can we follow up with a ticket to change the string for a localized label?

**Comment by gmajoulet** [comment]
Adding the translation key should be pretty straightforward, we have some documentation here: https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/1.0/_locales/README.md

It's slightly outdated in how to format the strings but you can just copy/paste what you see in the language files. :))

**Review by processprocess — APPROVED** [review]
LGTM. Thanks @nickFalcone :)

I'll mention this PR in the localization label PR so you can see how it's implemented if you're curious about it!

**Comment by processprocess** [comment]
Building off of @gmajoulet comment:
If you want to add the translation key in this PR I think that would be great.

Here is an example in the component of how to set the key in the HTML once the string is created:
https://github.com/ampproject/amphtml/blob/master/extensions/amp-story-360/0.1/amp-story-360.js#L599-L603

**Comment by nickFalcone** [comment]
I will incorporate the translation key into this PR.  Thanks for the examples and links @processprocess!

**Comment by nickFalcone** [comment]
Hi @processprocess, I added the translation key in 26bba5c 😄

File: extensions/amp-story/1.0/_locales/en.json [hunk 1]
```diff
@@ -250,5 +250,9 @@
   "89": {
     "string": "Your response will be sent to",
     "description": "Text displayed to users after clicking a button that reveals a disclaimer, telling them more about where their user data will be stored, after they interact with a poll or quiz."
+  },
+  "90": {
+    "string": "Move device to explore",
+    "description": "Text displayed to users on first instance of an AMP 360 Story, telling them to move their device to experience the gyroscope effect."
```
Comment (processprocess) [inline extensions/amp-story/1.0/_locales/en.json:256]:
```suggestion
    "description": "Text displayed to users on gyroscope activation of an amp-story-360 component, telling them to move their device to experience the gyroscope effect."
```

**Comment by processprocess** [comment]
Thanks @nickFalcone! Looking good. Minor nit with wording in the comment.

**Comment by processprocess** [comment]
Looks good @nickFalcone. Thank you!
cc @kristoferbaxter for owners approval.

**Comment by gmajoulet** [comment]
cc @erwinmombay for OWNERS approval

**Comment by nickFalcone** [comment]
Hi @erwinmombay, I just resolved a few conflicts in f208fea 😃

**Review by erwinmombay — APPROVED** [review]

File: extensions/amp-story/1.0/_locales/default.json [hunk 2]
```diff
@@ -38,5 +38,6 @@
   "84": {"string": "SCORE:"},
   "85": {"string": "Pause"},
   "86": {"string": "Play"},
-  "89": {"string": "Your response will be sent to"}
+  "89": {"string": "Your response will be sent to"},
+  "90": {"string": "Move device to explore"}
```
Comment (Enriqe) [inline extensions/amp-story/1.0/_locales/default.json:42]:
```suggestion
  "96": {"string": "Move device to explore"}
```

Could you update this string to reflect the appropriate key? I think we forgot to add a few here.

**Comment by nickFalcone** [comment]
Hi @ampproject/build-cop, is it possible to pass the failing [codecov/patch](https://github.com/ampproject/amphtml/pull/32964/checks?check_run_id=2135951686) test?

**Comment by Enriqe** [comment]
All good to merge @nickFalcone, that check is not blocking.

**Comment by nickFalcone** [comment]
Thanks @Enriqe! 🎉

**Comment by Enriqe** [comment]
Thank you @nickFalcone, great job!👏

**Pull request merged**